### DISCUSSION TITLE
Query regarding Correct Newline Detection in Parallel CSV Copy Parsing

### DISCUSSION THREAD

**Discussion by gvm001** [duckdb/duckdb#15468] [Q&A]
I have a doubt regarding how DuckDB's CSV parser handles escaped characters when processing a CSV file in chunks for parallel parsing. Specifically, I am curious about how the parser determines the starting index of rows when a chunk starts within an escape sequence inside a quoted field.

**Example CSV:**
Consider the following CSV content: [Table has two text columns]
```
"xx\"a\nb","b,c"
"a,b","c,d"
```

If the chunk starts after "xx\\" (at the position of **"** after escape \\"), the parser seems to misinterpret the escape sequence, leading to an incorrect start index for rows. Based on the new line finder algorithm:

In the STANDARD state, the parser reads until the next newline (\n). If the newline position matches the first newline in the buffer and the row is valid, it is accepted.
If the STANDARD state fails, the parser moves to the QUOTED state and attempts to parse from there.
If both STANDARD and QUOTED states fail, the parser switches to the ESCAPED state.
Observed Behavior:
In this case:

If the parser begins in the STANDARD state, it processes until it finds a newline (\n). Here, it interprets "a\n" as valid new line and assumes the row is "b","b,c". This seems incorrect.
If the parser begins in the QUOTED state, it still interprets "b","b,c" as valid, starting from b.
It is unclear whether the ESCAPED state is reached or how it resolves such cases.
**Questions:**
**Escape Handling:**
Does the parser consider the ESCAPED state when chunks begin within an escape sequence (e.g., \\")? If so, how is the start index of the row decided in such cases?

**State Transition Logic:**
If the STANDARD and QUOTED states both produce valid but incorrect rows, will the ESCAPED state take precedence? What criteria are used to decide this?

**Expected Parsing Behavior:**
For this example, should the parser treat \\" as part of the quoted sequence and correctly parse the row starting from "xx\\"a\nb","b,c"? How should such rows spanning chunks be handled?

Any guidance or clarification on how DuckDB handles such cases would be greatly appreciated.

**Comment by Tishj** [comment]
I believe Pedro is working on a paper about csv reading, I'm not entirely sure this specific detail will be covered in this one, that might be out of scope.
If it is, then it might be potentially covered in a future paper

For now all I can say is that it involves following multiple states. One of those will be treating the chunk as if we are within a quoted state, so two chunks can be properly resolved if the preceding chunk ended in a quoted state
I imagine ESCAPED is treated in a similar fashion
